felix/bpf/consistenthash/consistenthash.go

```diff
@@ -0,0 +1,191 @@
+//     of LUT positions it would like to occupy.
+//   - Constructs a ConsistentHash backend LUT to be hashed-into with packet 5-tuples.
+type ConsistentHash struct {
+	// m is a prime number.
+	// Defaults to ConsistentHash.M
+	m      int
+	h1, h2 hash.Hash
+
+	// Lexicographically orders the backends by name.
+	backendNames   []string
+	backendsByName map[string]backend
+}
+
+type backend struct {
+	permutation []int
+	endpoint    k8sp.Endpoint
+}
+
+// New returns a backend-hashing module.
+func New(lutSize int, hash1, hash2 hash.Hash) *ConsistentHash {
+	c := &ConsistentHash{m: lutSize}
+	c.backendNames = make([]string, 0)
+	c.backendsByName = make(map[string]backend)
+
+	if hash1 == nil || hash2 == nil {
+		panic("nil hashing function for ConsistentHash")
+	}
+	c.h1 = hash1
+	c.h2 = hash2
+
+	return c
+}
+
+// AddBackend generates and stores a permutation for the given backend name,
+// to be factored into the LUT generation.
+func (ch *ConsistentHash) AddBackend(kep k8sp.Endpoint) {
+	var b backend
+
+	if kep == nil {
+		logrus.Warn("Ignoring AddBackend for nil endpoint")
+		return
+	}
+
+	name := kep.String()
+	if _, exists := ch.backendsByName[name]; exists {
+		logrus.WithField("backend", name).Info("Will not regenerate permutation for pre-existing backend")
+		return
+	}
+
+	permutation, err := ch.permutation(name)
+	if err != nil {
+		logrus.WithError(err).WithField("backend", name).Error("Failed to generate permutation for backend")
+		return
+	}
+
+	b.endpoint = kep
+	b.permutation = permutation
+
+	ch.backendsByName[name] = b
+	ch.backendNames = append(ch.backendNames, name)
+}
+
+func (ch *ConsistentHash) RemoveBackend(kep k8sp.Endpoint) {
+	name := kep.String()
+	if kep == nil {
+		logrus.Warn("Ignoring RemoveBackend for nil endpoint")
+		return
+	}
+
+	b, exists := ch.backendsByName[name]
+	if !exists {
+		logrus.WithField("backend", name).Warn("Ignoring RemoveBackend for non-existent backend")
+		return
+	}
+
+	b.endpoint = nil
+	delete(ch.backendsByName, name)
+
+	for i, b := range ch.backendNames {
+		if b == name {
+			slices.Delete(ch.backendNames, i, i)
+		}
+	}
+}
+
+// Generate sorts the list of backends and then generates a ConsistentHash LUT.
+func (ch *ConsistentHash) Generate() []k8sp.Endpoint {
+	if len(ch.backendNames) == 0 {
+		return nil
+	}
+
+	slices.Sort(ch.backendNames)
+	logrus.WithField("backends", ch.backendNames).Info("sorted backend names")
+
+	// Next-preference for each backend.
+	next := make([]int, len(ch.backendNames))
+	// The final lookup-table to hash against.
+	lut := make([]k8sp.Endpoint, ch.m)
+
+	// In total, we go to M iterations of the inner loop.
+	// Can't rely on the outer-loop condition to break at the right time,
+	// so we're counting manually.
+	n := 0
+populate:
+	for {
+		for i, backend := range ch.backendNames {
+			prefs := ch.backendsByName[backend].permutation
+			choice := prefs[next[i]]
+			for lut[choice] != nil {
+				next[i]++
+				choice = prefs[next[i]]
```

Comment:
may you get out of bounds here? can next[i] get larger than len(prefs) ?

Reply:
`next[i]++` may technically get over. Isn't that some sort of a corner case a break-out-of-the-loop condition?